File: Frontend/src/api/pets.ts
import axios, { AxiosPromise } from 'axios'
import { BE_BASE_URL as URL } from '../constants'
import { CreatePet, Pet, UpdatePet } from '../types/pet'

export const getPets = (): AxiosPromise<{pets: Pet[]}> => {
    return axios.get(`${URL}/pets`)
}

export const createPet = (pet: CreatePet):AxiosPromise<{pet: Pet}>  => {
    const formData = new FormData()
    formData.append('name', pet.name)
    formData.append('type', pet.type)
    formData.append('gender', pet.gender)
    formData.append('country', pet.country)
    return axios.post(`${URL}/pets`, formData)
}

export const getPetProfile = (petId: number): AxiosPromise<{pet: Pet}> => {
    return axios.get(`${URL}/pets/${petId}`)
}

export const updatePet = (pet: UpdatePet): AxiosPromise<{ updated_pet: Pet }> => {
    const json = JSON.stringify(pet)
    return axios.put(`${URL}/pets/${pet.id}`, json)
}

export const deletePet = (petId: number): AxiosPromise<{deleted_pet: Pet}> => {
    return axios.delete(`${URL}/pets/${petId}`)
}

export const adoptPet = (petId: number, member_id: number): AxiosPromise<{ adopted_pet: Pet }> => {
    const json = JSON.stringify({ member_id })
    return axios.put(`${URL}/pets/${petId}/add`, json)
}

export const removePet = (petId: number): AxiosPromise<{ removed_pet: Pet }> => {
    return axios.put(`${URL}/pets/${petId}/remove`)
}

File: backend/application/views/pets.py
import json
from django.http import JsonResponse, HttpRequest
from ..models import Pets
from ..validators import CustomValidation
from django.db.models import F


def adopt_pet(request, pet_id):  # PUT
    data = json.loads(request.body)
    user_id = data["user_id"]
    pet = Pets.objects.get(id=pet_id)
    pet.user_id = user_id
    pet.save()
    return JsonResponse(
        {
            "adopted_pet": {
                "id": pet.pk,
                "name": pet.name,
                "type": pet.type,
                "gender": pet.gender,
                "user_id": pet.user_id,
                "country": pet.country,
            },
        }
    )


def remove_pet(request, pet_id):  # PUT
    pet = Pets.objects.get(id=pet_id)
    pet.user_id = None
    pet.save()
    return JsonResponse(
        {
            "removed_pet": {
                "id": pet.pk,
                "name": pet.name,
                "type": pet.type,
                "gender": pet.gender,
                "user_id": pet.user_id,
                "country": pet.country,
            },
        }
    )


def handle_pet(request: HttpRequest, id):

    if request.method == "GET":
        pet = Pets.objects.get(id=id)
        return JsonResponse(
            {
                "pet": {
                    "id": pet.pk,
                    "name": pet.name,
                    "type": pet.type,
                    "gender": pet.gender,
                    "user_id": pet.user_id,
                    "country": pet.country,
                },
            }
        )

    if request.method == "PUT":
        data = json.loads(request.body)
        validated = CustomValidation(
            {
                "name": {
                    "value": data["name"],
                    "validators": {"type": "str", "max_length": 10, "min_length": 3},
                },
                "type_field": {
                    "value": data["type"],
                    "validators": {"type": "str", "max_length": 8, "min_length": 3},
                },
                "gender": {
                    "value": data["gender"],
                    "validators": {"type": "str", "max_length": 10, "min_length": 3},
                },
            }
        )

        if validated.is_valid == False:
            return validated.get_errors()

        pet = Pets.objects.get(id=id)
        pet.name = data["name"]
        pet.type = data["type"]
        pet.gender = data["gender"]
        pet.country = data["country"]
        pet.save()
        return JsonResponse(
            {
                "updated_pet": {
                    "id": pet.pk,
                    "name": pet.name,
                    "type": pet.type,
                    "gender": pet.gender,
                    "user_id": pet.user_id,
                    "country": pet.country,
                },
            }
        )

    if request.method == "DELETE":
        pet = Pets.objects.get(id=id)
        pet.delete()
        return JsonResponse(
            {
                "deleted_pet": {
                    "id": pet.pk,
                    "name": pet.name,
                    "type": pet.type,
                    "gender": pet.gender,
                    "user_id": pet.user_id,
                    "country": pet.country,
                },
            }
        )


def handle_all_pets(request: HttpRequest):

    if request.method == "POST":
        data = json.loads(request.body)
        name = data["name"]
        type = data["type"]
        gender = data["gender"]
        country = data["country"]
        validated = CustomValidation(
            {
                "name": {
                    "value": name,
                    "validators": {"type": "str", "max_length": 25, "min_length": 1},
                },
                "typefield": {
                    "value": type,
                    "validators": {"type": "str", "max_length": 25, "min_length": 1},
                },
                "gender": {
                    "value": gender,
                    "validators": {"type": "str", "max_length": 10, "min_length": 3},
                },
            }
        )

        if validated.is_valid == False:
            return validated.get_errors()
        pet = Pets(name=name, type=type, gender=gender, country=country)
        pet.save()
        return JsonResponse(
            {
                "pet": {
                    "id": pet.pk,
                    "name": pet.name,
                    "type": pet.type,
                    "gender": pet.gender,
                    "country": pet.country,
                },
                "user_id": pet.user_id,
            }
        )

    if request.method == "GET":
        pets = Pets.objects.all().values()
        return JsonResponse(
            {
                "pets": list(pets),
            }
        )
